<title>Integration</title>

## Integration
by Eugenio Andrieu 

Integration is one of the basic operations in calculus. The [scipy.integrate](https://docs.scipy.org/doc/scipy/reference/integrate.html#module-scipy.integrate) sub-package provides an easy-to-use set of functions that allows users to calculate simple, double and tripple integrals. In this tutorial we will see how to use them to calculate the integrals of some simple functions.

## Simple integrals

Integrals of univariate functions can be done with the quad function. In its simplest form, this function has the following syntax:<br/><br/>
**<div align="center"> scipy.integrate.quad(func, a, b)</div>**<br/>
where
a and b are float values that represent the lower and upper limit of the integration respectively, and func is the function to be integrated.<br/>
In the example code, we will calculate the following integral:

<div align="center">$\int\limits_0^1$xdx = $^1/_2$</div>	

In [4]:
# Import the integrate subpackage
import scipy.integrate as integrate

# Define the function to be integrated. In our case, f(x) =  x.
def myFunction(x):
    return x
    
# Invoke the quad function and integrate myFunction between 0 and 1.
result = integrate.quad(myFunction, 0, 1)

# Show the result
result

(0.5, 5.551115123125783e-15)

As we can see, the results shows the correct value of 0.5 or 1/2, with a calculated error of 5e-15.

## Double integrals

Integrals of functions with two variables can be done with the dblquad function. In its simplest form, this function has the following syntax:


**<div align="center">scipy.integrate.dblquad(func, a, b,f(x),g(x))</div>**

where:<br/><br/>
a and b are float values that represent the lower and upper limit in x ,f(x) and g(x) are functions in x that represent the limits in y, and func is the function to be integrated.<br/>
In the example code, we will calculate the following integral:<br/>
<div align="center">$\int\limits_0^1$$\int\limits_0^1$ xy dxdy = $^1/_4$.</div>	

In [2]:
from scipy.integrate import dblquad

def integrand(y, x):  # Inverted order: y must be the first argument, and x the second. 
    return y * x

result, err = dblquad(integrand, 0, 1,
                   lambda x: 0, # we use the lambda function because the lower limit in y has to be a function in x
                   lambda x: 1) # we use the lambda function because the lower limit in y has to be a function in x
print(result) 
print(err)

0.24999999999999997
5.539061329123429e-15


As we can see, the results shows the correct value of 0.25 or 1/2, with a calculated error of 5e-15.

Integrals of functions with three variables can be done with the quad function. In its simplest form, this function has the following syntax:


**<div align="center">scipy.integrate.tplquad(func, a, b,f(x),g(x),h(x,y),i(x,y))</div>**

where:<br/><br/>
a and b are float values that represent the lower and upper limit in x , f(x) and g(x) are functions in x that represent the limits in y, h(x,y) and i(x,y) are functions in x and y that represent the limits in z, and func is the function to be integrated.<br/>
In the example code, we will calculate the following integral:<br/>
<div align="center">$\int\limits_0^1$$\int\limits_0^1$$\int\limits_0^1$ xyz dxdy = $^1/_8$.</div>	

In [3]:
from scipy.integrate import tplquad

def integrand(z, y, x):  # Inverted order: z must be the first argument, y the second and x the third. f(x,y,z)
    return y * x * z

result, err = tplquad(integrand, 0, 1,
                   lambda x: 0, # f(x) = 0. We use the lambda function because the lower limit in y has to be a function in x
                   lambda x: 1, # g(x) = 1
                   lambda x,y: 0, # h(x,y) = 0
                   lambda x,y: 1)  # i(x,y) = 1
                 
print(result) 
print(err)

0.12499999999999999
5.527033708952211e-15


As we can see, the results shows the correct value of 0.125 or 1/2, with a calculated error of 6e-15.